tests/integration/events.go

```diff
@@ -148,52 +150,76 @@ func waitForUnsubscribeToCollectionEvent(s *state, action UnsubscribeToCollectio
 // update event to the local event bus.
 //
 // Expected document heads will be updated for any connected nodes.
-func waitForUpdateEvents(s *state, nodeID immutable.Option[int], collectionID int) {
+func waitForUpdateEvents(
+	s *state,
+	nodeID immutable.Option[int],
+	updates map[string]struct{},
+) {
 	for i := 0; i < len(s.nodes); i++ {
 		if nodeID.HasValue() && nodeID.Value() != i {
 			continue // node is not selected
 		}
 
-		var evt event.Update
-		select {
-		case msg, ok := <-s.nodeEvents[i].update.Message():
-			if !ok {
-				require.Fail(s.t, "subscription closed waiting for update event")
-			}
-			evt = msg.Data.(event.Update)
-
-		case <-time.After(eventTimeout):
-			require.Fail(s.t, "timeout waiting for update event")
+		expect := make(map[string]struct{})
+		for k := range updates {
+			expect[k] = struct{}{}
 		}
 
-		if i >= len(s.nodeConfigs) {
-			return // not testing network state
-		}
+		for len(expect) > 0 {
+			var evt event.Update
+			select {
+			case msg, ok := <-s.nodeEvents[i].update.Message():
+				if !ok {
+					require.Fail(s.t, "subscription closed waiting for update event")
+				}
+				evt = msg.Data.(event.Update)
 
-		// make sure the event is published on the network before proceeding
-		// this prevents nodes from missing messages that are sent before
-		// subscriptions are setup
-		time.Sleep(100 * time.Millisecond)
+			case <-time.After(eventTimeout):
+				require.Fail(s.t, "timeout waiting for update event")
+			}
 
-		// update the actual document head on the node that updated it
-		s.nodeP2P[i].actualDocHeads[evt.DocID] = evt.Cid
+			// make sure the event is expected
+			_, ok := expect[evt.DocID]
+			require.True(s.t, ok, "unexpected document update")
+			delete(expect, evt.DocID)
 
-		// update the expected document heads of replicator targets
-		for id := range s.nodeP2P[i].replicators {
-			// replicator target nodes push updates to source nodes
-			s.nodeP2P[id].expectedDocHeads[evt.DocID] = evt.Cid
-		}
+			if i >= len(s.nodeConfigs) {
+				continue // not testing network state
```

Comment:
todo: This comment caused me a fair bit of confusion, and I initially thought it was skipping P2P tests (via some strange mechanic that I didn't yet understand).  Can you please rework the comment a bit?

Reply:
I updated the docs and broke out the rest into a more dedicated function. Hopefully its more clear now.